lib/TestRunner.lua

```diff
@@ -53,51 +102,36 @@ function TestRunner.runPlanNode(session, planNode, tryStack, noXpcall)
 				childResultNode.status = TestEnum.TestStatus.Skipped
 			else
 				if tryStack:size() > 0 and tryStack:getBack().isOk == false then
+
 					childResultNode.status = TestEnum.TestStatus.Failure
 					table.insert(childResultNode.errors,
 						string.format("%q failed without trying, because test case %q failed",
 							childPlanNode.phrase, tryStack:getBack().failedNode.phrase))
 				else
 					-- Errors can be set either via `error` propagating upwards or
 					-- by a test calling fail([message]).
-					local success = true
-					local errorMessage
-
-					local testEnvironment = getfenv(childPlanNode.callback)
+					local success, errorMessage = true, nil
 
-					for key, value in pairs(TestRunner.environment) do
-						testEnvironment[key] = value
+					for _, hook in pairs(lifecycleHooks:getPendingBeforeHooks()) do
+						if success then
+							success, errorMessage = runCallback(hook, false, "beforeAll hook: ")
+						end
 					end
 
-					testEnvironment.fail = function(message)
-						if message == nil then
-							message = "fail() was called."
+					for _, hook in pairs(lifecycleHooks:getBeforeEachHooks()) do
+						if success then
+							success, errorMessage = runCallback(hook, false, "beforeEach hook: ")
 						end
-
-						success = false
-						errorMessage = message .. "\n" .. debug.traceback()
 					end
 
-					-- We prefer xpcall, but yielding doesn't work from xpcall.
-					-- As a workaround, you can mark nodes as "not xpcallable"
-					local call = noXpcall and pcall or xpcall
-
-					-- Any code can check RUNNING_GLOBAL to fork behavior based on
-					-- whether a test is running. We use this to avoid accessing
-					-- protected APIs; it's a workaround that will go away someday.
-					_G[RUNNING_GLOBAL] = true
-
-					local nodeSuccess, nodeResult = call(childPlanNode.callback, function(message)
-						return message .. "\n" .. debug.traceback()
-					end)
-
-					_G[RUNNING_GLOBAL] = nil
+					if success then
+						success, errorMessage = runCallback(childPlanNode.callback)
+					end
 
-					-- If a node threw an error, we prefer to use that message over
-					-- one created by fail() if it was set.
-					if not nodeSuccess then
-						success = false
-						errorMessage = nodeResult
+					for _, hook in pairs(lifecycleHooks:getAfterEachHooks()) do
+						if success then
+							success, errorMessage = runCallback(hook, true, "afterEach hook: ")
+						end
```

Comment:
We have to branch on `success` a lot here to make sure we aren't going to continue execution if one of our steps fails. Do you think it would be a good idea to refactor this block into a function and use early returns instead?